pymc/tests/logprob/test_joint_logprob.py

```diff
@@ -309,3 +317,279 @@ def test_multiple_rvs_to_same_value_raises():
+
+    for i in range(10):
+        bern_sp = sp.bernoulli(p)
+        I_value = bern_sp.rvs(size=size).astype(I_rv.dtype)
+
+        norm_sp = sp.norm(mu[I_value, np.ogrid[mu.shape[1] :]], sigma)
+        A_idx_value = norm_sp.rvs().astype(A_idx.dtype)
+
+        exp_obs_logps = norm_sp.logpdf(A_idx_value)
+        exp_obs_logps += bern_sp.logpmf(I_value)
+
+        logp_vals = logp_vals_fn(A_idx_value, I_value)
+
+        np.testing.assert_almost_equal(logp_vals, exp_obs_logps, decimal=decimals)
+
+
+def test_logp_helper():
+    value = at.vector("value")
+    x = pm.Normal.dist(0, 1)
+
+    x_logp = pm.logp(x, value)
+    np.testing.assert_almost_equal(x_logp.eval({value: [0, 1]}), sp.norm(0, 1).logpdf([0, 1]))
+
+    x_logp = pm.logp(x, [0, 1])
+    np.testing.assert_almost_equal(x_logp.eval(), sp.norm(0, 1).logpdf([0, 1]))
+
+
+def test_logp_helper_derived_rv():
+    assert np.isclose(
+        pm.logp(at.exp(pm.Normal.dist()), 5).eval(),
+        pm.logp(pm.LogNormal.dist(), 5).eval(),
+    )
+
+
+def test_logp_helper_exceptions():
+    with pytest.raises(TypeError, match="When RV is not a pure distribution"):
+        pm.logp(at.exp(pm.Normal.dist()), [1, 2])
+
+    with pytest.raises(NotImplementedError, match="PyMC could not infer logp of input variable"):
+        pm.logp(at.cos(pm.Normal.dist()), 1)
+
+
+def test_model_unchanged_logprob_access():
+    # Issue #5007
+    with pm.Model() as model:
+        a = pm.Normal("a")
+        c = pm.Uniform("c", lower=a - 1, upper=1)
+
+    original_inputs = set(pytensor.graph.graph_inputs([c]))
+    # Extract model.logp
+    model.logp()
+    new_inputs = set(pytensor.graph.graph_inputs([c]))
+    assert original_inputs == new_inputs
+
+
+def test_unexpected_rvs():
+    with pm.Model() as model:
+        x = pm.Normal("x")
+        y = pm.CustomDist("y", logp=lambda *args: x)
+
+    with pytest.raises(ValueError, match="^Random variables detected in the logp graph"):
+        model.logp()
+
+
+def test_hierarchical_logp():
+    """Make sure there are no random variables in a model's log-likelihood graph."""
+    with pm.Model() as m:
+        x = pm.Uniform("x", lower=0, upper=1)
+        y = pm.Uniform("y", lower=0, upper=x)
+
+    logp_ancestors = list(ancestors([m.logp()]))
+    ops = {a.owner.op for a in logp_ancestors if a.owner}
+    assert len(ops) > 0
+    assert not any(isinstance(o, RandomVariable) for o in ops)
+    assert m.rvs_to_values[x] in logp_ancestors
+    assert m.rvs_to_values[y] in logp_ancestors
+
+
+def test_hierarchical_obs_logp():
+    obs = np.array([0.5, 0.4, 5, 2])
+
+    with pm.Model() as model:
+        x = pm.Uniform("x", 0, 1, observed=obs)
+        pm.Uniform("y", x, 2, observed=obs)
+
+    logp_ancestors = list(ancestors([model.logp()]))
+    ops = {a.owner.op for a in logp_ancestors if a.owner}
+    assert len(ops) > 0
+    assert not any(isinstance(o, RandomVariable) for o in ops)
+
+
+def test_logprob_join_constant_shapes():
```

Comment:
This should be in another file, corresponding to where the rewrite is applied. Also we can remove the aeppl comment below

Reply:
This test is redundant of https://github.com/pymc-devs/pymc/blob/c5e449709adeab3e15aa0b4243efb454fa599bf8/pymc/tests/logprob/test_tensor.py#L218-L326

EXCEPT the line `assert_no_rvs(...)` on the logp expression. We should add that line in those tests